config/src/config.rs

```diff
@@ -125,7 +125,7 @@ impl ConsensusPrivateKey {
                 let key = std::mem::replace(self, ConsensusPrivateKey::Removed);
                 match key {
                     ConsensusPrivateKey::Present(priv_key) => Some(priv_key),
-                    _ => unreachable!(),
+                    _ => None,
```

Comment:
@mimoo: unreachable!() is like an assert here, since we really should never reach this line after the check on line 125. `None` means that this path is OK to take. I'd argue in favor of keeping the panic.

Reply:
Let me ask you this question: is returning `None` here instead of panicking a problem? Will this have further potential consequences to the flow?

Reply:
`None` could be confused with the key not being present, as on line 131 in the same file. This could cause an "action at a distance" phenomenon if for some reason the `mem::replace` on line 125 does not return what we would expect it to return.

Reply:
I would argue that this is why we have `Option`, it forces the caller to handle the `None` case.  If this confuses the caller, then there might be a bug in the caller?

Eventually this is your call, and indeed this should never be reached.

(I'm on a quest to remove all `unreachable` to prompt people to start removing them :P)

Reply:
>If this confuses the caller, then there might be a bug in the caller?

The caller is going to handle the `None` case as if the key didn't exist, and there's not much else it can do. I wouldn't call that a bug. If we reach the `unreachable!()` arm, the bug is really in `mem::replace` and IMO catching the error immediately is better than letting it propagate.

More philosophically, `unreachable!()` is like `assert!()` but for control flow - why are we looking to remove them?

Reply:
The current code guidelines suggest that unreachable!() is fine.
https://github.com/libra/libra/blob/master/documentation/coding_guidelines.md#error-handling

However in this case, we handle this code somewhat inconsistently from other parts of the codebase (e.g. in the same file see take_network_signing_private()) - this doesn't worry about the unreachable() and instead calls expect() later.  It would be good to unify on the style.

Reply:
> More philosophically, unreachable!() is like assert!() but for control flow - why are we looking to remove them?

here's my reasoning: we're building high-assurance software, this means we should never panic unless we can't recover from it. The problem with unreachable is that they are assumptions from the developer, if the developer made a mistake then this can lead to a crash. This is actually what caused our very first crash in prod!

Worse, our static analyzer (MIRAI) will see `unreachable!()` macros as assumptions as well, and will not verify them (whereas it will do its best to make sure that `debug_assert!(false)` or `assert!(false)` are never reached (cf @hermanventer)).

I would argue that we should modify our current guidelines to never use unreachable, and use either `assert!(false)` or `debug_assert!(false)` instead so that at least MIRAI can check these assumptions.

Reply:
actually, re-reading our coding guidelines I just think we should make an effort to define "unrecoverable errors" and to make sure to add a comment whenever there is a function that panics to explain why we think it's necessary and why it's unrecoverable.

Reply:
If an error happens because code has been incorrectly written, it is unrecoverable because nothing makes sense anymore and the only reasonable thing is to restart execution from a safe point. If it happens for another reason, such as bad user input, it should always be recoverable.

Our goal is to statically ensure that no errors will ever happen because code has been incorrectly written. MIRAI is the tool that we use to verify this. For various reasons, there are some things that MIRAI cannot verify, but on which we need to rely upon anyway, presumably for good reasons that are both documented and well understood by code reviewers. One way to expression such an assumption to MIRAI is via unreachable! (because of lot of existing code uses unreachable! this way).

In our own code, we should provide convincing comments justifying why we use unreachable!, just as we would for assume! annotations. And we should be VERY sure that the code will never be reached.

Sometimes this convincing comment will be in the form a MIRAI annotation that can in fact be proved. In that case the unreachable! may be removed if the type checker allows it. If not, just keep it.

Reply:
Thanks for chiming in, @hermanventer.

> If an error happens because code has been incorrectly written, it is unrecoverable because nothing makes sense anymore and the only reasonable thing is to restart execution from a safe point. If it happens for another reason, such as bad user input, it should always be recoverable.

What about the case where it happens due to system errors, such as running out of disk or memory? It could be argued that it is not buggy code, but a violation of basic assumptions in the application code.
 
> Our goal is to statically ensure that no errors will ever happen because code has been incorrectly written. MIRAI is the tool that we use to verify this. For various reasons, there are some things that MIRAI cannot verify, but on which we need to rely upon anyway, presumably for good reasons that are both documented and well understood by code reviewers. One way to expression such an assumption to MIRAI is via unreachable! (because of lot of existing code uses unreachable! this way).
> 
> In our own code, we should provide convincing comments justifying why we use unreachable!, just as we would for assume! annotations. And we should be VERY sure that the code will never be reached.
> 

This is a good point. I'll add comments on why it's OK to crash if `mem::replace` is behaving unexpectedly.

> Sometimes this convincing comment will be in the form a MIRAI annotation that can in fact be proved. In that case the unreachable! may be removed if the type checker allows it. If not, just keep it.

I am not familiar with MIRAI, but my guess is that since the Rust standard library is not annotated it would be hard to "prove" an assertion that mem::replace should behave a certain way. Please correct me if I'm wrong here.

Reply:
> This is a good point. I'll add comments on why it's OK to crash if mem::replace is behaving unexpectedly.

`mem::replace` behaving unexpectedly might be a good reason to crash indeed! 

if you do this, can you add a comment right above the panic that starts with `// unrecoverable: blablabla`. I'm thinking of writing a tool that would catch any panic that doesn't have such a comment :P

Reply:
> What about the case where it happens due to system errors, such as running out of disk or memory? It could be argued that it is not buggy code, but a violation of basic assumptions in the application code.

Arguably, running a program on a system that runs out of disk or memory is a programming mistake. A well written service protects against this, programmatically.

> the Rust standard library is not annotated 

Indeed. There is, however, a way to provide such annotations. MIRAI can also infer annotations. So all is not lost and this could well be do-able.

Meanwhile, unreachable!() with a comment saying why this is unreachable seems like a good way to proceed.